### PR TITLE
Extend mz_audit_events.details with a reason that shows which REFRESH MV prompted turning on a cluster

### PR DESCRIPTION
**Pull request by ggevay**
As [mentioned in the REFRESH design doc](https://github.com/ggevay/materialize/blob/refresh-every-design-doc/doc/developer/design/20231027_refresh_mvs.md#introspection--observability), this extends `mz_audit_events.details` with a reason that shows which REFRESH MV prompted turning on a cluster. Two audit log event types are involved: replica creates and drops. Replica creates are getting
```
"scheduling_decision_reasons":[{"RefreshTurnOn":{"mvs_needing_refresh":["u42", "u43"]}}]
```
while replica drops are getting
```
"scheduling_decision_reasons":["RefreshTurnOff"]
```
Note that replica creates and drops that are _not_ initiated by the auto scheduler are getting
```
"scheduling_decision_reasons":null
```
where `null` is json's null.

In the future, I expect `scheduling_decision_reasons` to have more possible values when we'll have more cluster scheduling policies.

Note that the catalog migration is a no-op, because I'm not actually migrating old audit log events, but instead created a V2 of `CreateClusterReplica` and `DropClusterReplica`. This was suggested by Joe [here](https://materializeinc.slack.com/archives/C02FWJ94HME/p1716383610709829?thread_ts=1716381334.178889&cid=C02FWJ94HME).

Tagging @ParkMyCar for review, because he reviewed earlier cluster scheduling PRs, and also recently modified drop catalog operations, which I'm also touching here.

### Motivation

  * This PR adds a known-desirable feature: https://github.com/MaterializeInc/materialize/issues/27072

### Tips for reviewer

Commits:
1. The first commit is just a minor test change, slightly increasing some sleeps in REFRESH tests that were sometimes failing for me locally when ran with a busy CPU.
2. The second commit is the main thing.
3. The third commit addresses most of Nikhil's comments on how to present the information to the user.
4. The fourth commit changes how we stringify `Duration` based on [Nikhil's comment](https://github.com/MaterializeInc/materialize/pull/27199#issuecomment-2133599601), and it also makes sure in planning that this `Duration` is always convertible back to an `Interval` by actually trying to make that conversion.

I would suggest starting with the 1. and 4. commits separately, because these are small, and then just looking at the changes from all commits together. A top-down review could start from `cluster_scheduling.rs`.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered. ([trigger-ci for additional test/nightly runs](https://trigger-ci.dev.materialize.com/))
  - [I kicked off Nightly](https://buildkite.com/materialize/nightly/builds/7839), mostly for the upgrade tests. Btw., @MaterializeInc/testing, do you think we have such an upgrade test that involves a `SCHEDULE ON REFRESH` cluster and enough time on the old version for relevant audit log events to appear? If not, how could I go around to adding it? (I tested an upgrade scenario manually, by starting a local materialize from `main`, doing stuff that made relevant audit log events appear, and then restarting on my branch.)
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

File: src/adapter/src/coord/cluster_scheduling.rs [hunk 1]
```diff
@@ -72,12 +111,10 @@ impl Coordinator {
                             })
                             .collect_vec();
                         debug!(%cluster.id, ?refresh_mv_write_frontiers, "check_refresh_policy");
-                        min_refresh_mv_write_frontiers.push((
+                        refresh_mv_write_frontiers.push((
                             cluster.id,
                             rehydration_time_estimate,
-                            refresh_mv_write_frontiers
-                                .into_iter()
-                                .fold(Antichain::new(), |ac1, ac2| Lattice::meet(&ac1, &ac2)),
```
Comment (ggevay) [inline src/adapter/src/coord/cluster_scheduling.rs:80]: Moving this logic to the second half of this fucntion.

File: src/audit-log/src/lib.rs [hunk 2]
```diff
@@ -230,6 +232,17 @@ pub struct DropClusterReplicaV1 {
     pub replica_name: String,
 }
 
+#[derive(Clone, Debug, Serialize, Deserialize, PartialOrd, PartialEq, Eq, Ord, Hash, Arbitrary)]
+pub struct DropClusterReplicaV2 {
+    pub cluster_id: String,
+    pub cluster_name: String,
+    // Events that predate v0.32.0 will not have this field set.
+    #[serde(skip_serializing_if = "Option::is_none")]
```
Comment (ggevay) [inline src/audit-log/src/lib.rs:240]: I copy-pasted this from V1, but I guess it's no longer relevant here. Should I remove this?

File: src/audit-log/src/lib.rs [hunk 3]
```diff
@@ -244,6 +257,35 @@ pub struct CreateClusterReplicaV1 {
     pub internal: bool,
 }
 
+#[derive(Clone, Debug, Serialize, Deserialize, PartialOrd, PartialEq, Eq, Ord, Hash, Arbitrary)]
+pub struct CreateClusterReplicaV2 {
+    pub cluster_id: String,
+    pub cluster_name: String,
+    // Events that predate v0.32.0 will not have this field set.
+    #[serde(skip_serializing_if = "Option::is_none")]
```
Comment (ggevay) [inline src/audit-log/src/lib.rs:265]: (same here)

File: src/catalog/tests/snapshots/open__initial_audit_log.snap [hunk 4]
```diff
@@ -1,5 +1,6 @@
 ---
 source: src/catalog/tests/open.rs
+assertion_line: 444
```
Comment (ggevay) [inline src/catalog/tests/snapshots/open__initial_audit_log.snap:3]: @jkosh44, could you tell me what this line is? I think it appeared when I ran `test_persist_open()` / `test_persist_debug()`.

Comment (jkosh44) [reply src/catalog/tests/snapshots/open__initial_audit_log.snap:3]: Not positive, but it looks like the line in the file `open.rs` that asserts that this snapshot hasn't changed. It's most likely used to help with debugging output and is maybe a new feature of `insta`. This is all speculation, but other `.snap` files have it, so I wouldn't worry about it.

Comment (jkosh44) [reply src/catalog/tests/snapshots/open__initial_audit_log.snap:3]: Yeah, looks like it's just metadata about this snapshot, https://docs.rs/insta/latest/insta/internals/struct.MetaData.html#method.assertion_line.

**Comment by benesch** [comment]
I think the essence of the information included here is good, but the presentation looks a bit too much like debugging output, rather than something that's been tailored to be part of the public interface.

```json
// Turn on.
{"scheduling_decision_reasons":[{"RefreshTurnOn":{"mvs_needing_refresh":["u42", "u43"]}}]}

// Turn off.
{"scheduling_decision_reasons":["RefreshTurnOff"]}

// Manual.
{"scheduling_decision_reasons":null}
```

Here's an alternative formatting, for discussion:

```json
// Turn on.
{"reason": "schedule", "scheduling_policies": [{"type": "on-refresh", "eligible_objects": ["u42", "u43"], "rehydration_time_estimate": "5m"}]}

// Turn off.
{"reason": "schedule", "scheduling_policies": [{"type": "on-refresh", "eligible_objects": [], "rehydration_time_estimate": "5m"}]}

// Manual.
{"reason": "manual", "scheduling_policies": []}
```

It's late so I don't have time to write up more detailed thoughts, but wanted to at least get the idea out there.

**Comment by nrainer-materialize** [comment]
Thanks for triggering a nightly build! I suspect both failures could be flakes and triggered retries accordingly.

**Comment by ggevay** [comment]
> alternative formatting

In the alternative formatting, I like the `"reason": "schedule"` and `"reason": "manual"`, but the `"scheduling_policies"` might be too flat, because I can imagine different scheduling policies wanting to expose different fields. So how about further exploiting that we are in json and stuffing the rest of the fields below the `"type": "on-refresh"` as a nested json object?:
```
"scheduling_policies": [{"type": "on-refresh":{"eligible_objects": ["u42", "u43"], "rehydration_time_estimate": "50m"}}]
```
Also, won't it be confusing for users if the replica drop has the same fields as the replica create? Maybe they'll think that something is wrong when they see an empty list of `"eligible_objects": []`.

**Comment by nrainer-materialize** [comment]
The [coverage build](https://buildkite.com/materialize/coverage/builds/415) looks good!

**Comment by benesch** [comment]
```
"scheduling_policies": [{"type": "on-refresh":{"eligible_objects": ["u42", "u43"], "rehydration_time_estimate": "50m"}}]
```

@ggevay – this doesn't look like valid JSON to me!

> I can imagine different scheduling policies wanting to expose different fields

I was envisioning that we'd use serde's ["internally tagged"](https://serde.rs/enum-representations.html#internally-tagged) enum variant! For example:

```json
{"type": "on-refresh", "eligible_objects": []}
{"type": "fixed", "start": "09:00", "end": "11:00"}
```

The original format use had uses serde's "externally tagged" format. This is Serde's default, but I think it is quite a bit less natural in JSON/JavaScript. The "internally tagged" format, anecdotally, is much more common in the JSON/JavaScript that I've seen.

> Also, won't it be confusing for users if the replica drop has the same fields as the replica create? Maybe they'll think that something is wrong when they see an empty list of "eligible_objects": [].

Maybe it's better to use your original phrasing? `objects_needing_refresh: []` hopefully makes it clear that things were a success. My fear with having separate `RefreshTurnOn` and `RefreshTurnOff` reasons is that it doesn't reflect the structure of the underlying policy engine, which only answers the question "did this policy say turn on or not?". Seems more future proof to hew as closely to that policy engine as we can.

**Comment by benesch** [comment]
> @ggevay – this doesn't look like valid JSON to me!

Hm, on reflection, did you intend something like this?

```json
"scheduling_policies": {"on-refresh":{"eligible_objects": ["u42", "u43"], "rehydration_time_estimate": "50m"}}
```

With multiple active policies:

```json
"scheduling_policies": {"on-refresh":{"eligible_objects": ["u42", "u43"], "rehydration_time_estimate": "50m"}, "fixed": {"start_time": "0900", "end_time": "1100"}}
```

That seems like maybe our best option yet. After all, there can only be one decision per named policy, right?

**Comment by ggevay** [comment]
> Hm, on reflection, did you intend something like this?

Yes, exactly! Sorry for messing up my json example.

> After all, there can only be one decision per named policy, right?

Yes.

> That seems like maybe our best option yet.

Ok, then I'll change the code to this. So, to sum up the changes compared to the original PR:
- Add `rehydration_time_estimate`.
- Create a separate `"reason": "manual" / "schedule"`, and change `"scheduling_decision_reasons":null` to `"scheduling_policies": []`.
- No separate `RefreshTurnOn` and `RefreshTurnOff`, but just `on-refresh`, to make the mapping from scheduling policies 1-to-1.
- (We stay externally tagged.)

**Comment by benesch** [comment]
> * No separate `RefreshTurnOn` and `RefreshTurnOff`, but just `on-refresh`, to make the mapping from scheduling policies 1-to-1.

I _think_ so! But this is the one I'm least confident on. It may be helpful to look one step ahead. What's the next scheduling policy we expect to write, and how would we render that scheduling policy's decisions in the audit log. I feel like a fixed schedule is a good one to think through, since it's quite simple.

> * (We stay externally tagged.)

Just to confirm: this has no impact, right? Because we won't be serializing enums after your change?

**Comment by ggevay** [comment]
>> (We stay externally tagged.)

> Just to confirm: this has no impact, right? Because we won't be serializing enums after your change?

Hmm, I was confused, for some reason I thought the different scheduling policies would be captured in an enum. But I realize now that you probably meant `scheduling_policies` to be a struct that simply has one field for each existing policy. This seems fine, because each policy _must_ have an (On/Off + reason) opinion on each cluster.

>> No separate RefreshTurnOn and RefreshTurnOff, but just on-refresh, to make the mapping from scheduling policies 1-to-1.

> I think so! But this is the one I'm least confident on. It may be helpful to look one step ahead. What's the next scheduling policy we expect to write, and how would we render that scheduling policy's decisions in the audit log. I feel like a fixed schedule is a good one to think through, since it's quite simple.

Actually, we now have the problem that it's not clear whether a policy is saying that it wants to turn a cluster on or off. So how about adding a `decision` field?
```
"scheduling_policies": {
  "on-refresh":{
    "eligible_objects": ["u42", "u43"],
    "rehydration_time_estimate": "50m",
    "decision": "On"
  },
  "fixed": {
    "start_time": "0900",
    "end_time": "1100"
    "decision": "Off"
  }
}
```

**Comment by ggevay** [comment]
Also, if `"scheduling_policies"` is not a list of enum instances, but just a struct, then it should be `null` when the `reason` is `manual`, right?

**Comment by ggevay** [comment]
Thinking about one more policy example:
```
"scheduling_policies": {
  "on-refresh":{
    "eligible_objects": ["u42", "u43"],
    "rehydration_time_estimate": "50m",
    "decision": "On"
  },
  "fixed": {
    "start_time": "0900",
    "end_time": "1100",
    "decision": "Off"
  },
  "on-select": {
    "query_id": xxxxxxxxx, // that prompted the turn-on
    "last_select": "2024-05-24 15:40:00" // time of the last SELECT against the cluster
    "grace_period": "10m", // how long we keep a cluster On after a SELECT is done if there are no new SELECTs
    "decision": "Off",
  }
}
```

**Comment by ggevay** [comment]
I've done the above changes. Let me know what you think of the new format @benesch.
```
 id | event_type |    object_type    |                                                                                                                                         details                                                                                                                                          |    user     |        occurred_at         
----+------------+-------------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-------------+----------------------------
 21 | drop       | cluster-replica   | {"cluster_id":"u2","cluster_name":"c1","reason":"Schedule","replica_id":"u2","replica_name":"r1","scheduling_policies":{"on_refresh":{"decision":"Off","mvs_needing_refresh":[],"rehydration_time_estimate":"2s"}}}                                                                      |             | 2024-05-24 16:41:09.404+00
 20 | create     | cluster-replica   | {"billed_as":null,"cluster_id":"u2","cluster_name":"c1","disk":false,"internal":false,"logical_size":"1","reason":"Schedule","replica_id":"u2","replica_name":"r1","scheduling_policies":{"on_refresh":{"decision":"On","mvs_needing_refresh":["u2"],"rehydration_time_estimate":"2s"}}} |             | 2024-05-24 16:41:06.464+00
```

**Comment by ggevay** [comment]
Rebased.

New Nightly run: https://buildkite.com/materialize/nightly/builds/7874

**Comment by benesch** [comment]
I think the structure looks great! I really like how you've structured the `"decision": "on|off"` field, too.

Just a few style notes from [system catalog style](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/style.md#system-catalog-style):

  * Enum values should use kebab-case, so `"schedule"` and `"manual"` not `"Schedule"` and `"Manual"`. Similarly for `"on"` and `"off"`. Serde has an easy config option for this.
  * We should avoid all but the most common abbreviations. "mv" isn't an abbreviation that we used for materialized views elsewhere. I'd say just spell it out as `materialized_views_needing_refresh`, or `objects_needing_refresh` if we think things besides materialized views will need refreshes in the future.

> Also, if `"scheduling_policies"` is not a list of enum instances, but just a struct, then it should be `null` when the `reason` is `manual`, right?

Where did you land on this? I can see the argument for either `null` or `{}`.

**Comment by ggevay** [comment]
Ok, thanks! I'll fix the style things tomorrow.

>> Also, if "scheduling_policies" is not a list of enum instances, but just a struct, then it should be null when the reason is manual, right?

> Where did you land on this? I can see the argument for either null or {}.

It's currently `null`. I think I'd need a custom serializer to serialize None to `{}` instead (unless there is also some tricky annotation for this that I haven't found).

**Comment by benesch** [comment]
> It's currently null. I think I'd need a custom serializer to serialize None to {} instead (unless there is also some tricky annotation for this that I haven't found).

Maybe we should just omit the `scheduling_policies` key entirely when the schedule is `manual`? I think that's easier to set up with Serde (`skip_serializing_if`).

**Comment by ggevay** [comment]
Ah yes, that sounds good!

File: src/catalog/tests/snapshots/open__initial_audit_log.snap (hunk 4 above)
Comment (ggevay) [reply src/catalog/tests/snapshots/open__initial_audit_log.snap:3]: Ok, thank you!

**Comment by ggevay** [comment]
I've addressed the remaining comments about the format. Now it is:
```
 95 | drop       | cluster-replica   | {"cluster_id":"u2","cluster_name":"c1","reason":"schedule","replica_id":"u39","replica_name":"r1","scheduling_policies":{"on_refresh":{"decision":"off","objects_needing_refresh":[],"rehydration_time_estimate":"2s"}}}                                                                      |             | 2024-05-27 08:24:16.459+00
 94 | create     | cluster-replica   | {"billed_as":null,"cluster_id":"u2","cluster_name":"c1","disk":false,"internal":false,"logical_size":"1","reason":"schedule","replica_id":"u39","replica_name":"r1","scheduling_policies":{"on_refresh":{"decision":"on","objects_needing_refresh":["u2"],"rehydration_time_estimate":"2s"}}} |             | 2024-05-27 08:24:13.508+00
```

Note that even though the style guide says "Use the `interval` type to represent durations.", I'm thinking that this rule doesn't necessarily apply to the audit log's `details` field, because it is in json: I decided to represent `rehydration_time_estimate` by just a string (e.g. `40s`), which is the `Debug` representation of `Duration`. This is because if I were using the `Interval` type, then the json representation would be something like `{months: 0, days: 0, micros: 40000000}`, which seems less helpful than `40s`. Note that the same information is available from `mz_internal.mz_cluster_schedules`, where it _is_ represented as an `Interval`, so if somebody needs to do SQL operations on the rehydration time estimate, then they can use that one.

@ParkMyCar, could you please review the code part of the PR? I'd like to merge this soon, before I get one more big merge conflict.

**Comment by benesch** [comment]
> Note that even though the style guide says "Use the `interval` type to represent durations.", I'm thinking that this rule doesn't necessarily apply to the audit log's `details` field, because it is in json

Agreed. We need to do something slightly different. 

> I decided to represent `rehydration_time_estimate` by just a string (e.g. `40s`), which is the `Debug` representation of `Duration`. This is because if I were using the `Interval` type, then the json representation would be something like `{months: 0, days: 0, micros: 40000000}`, which seems less helpful than `40s`

Using a string is great, but we should use a string that's subject to our backwards compatibility guarantees—i.e., use our own string representation of the `interval` type. It'll be a little bit more typing, but then the format won't change if e.g. a new version of Rust changes the debug format of the `Duration` type. `mz_repr::strconv::format_interval(<code to make an interval from the duration you have>)`.

**Comment by ggevay** [comment]
> Using a string is great, but we should use a string that's subject to our backwards compatibility guarantees—i.e., use our own string representation of the interval type.

Done.

**Comment by benesch** [comment]
The format looks gorgeous! Thanks very much for iterating with me, Gabor. (Leaving the review of the actual code to more qualified folks.)

File: src/adapter/src/coord/cluster_scheduling.rs [hunk 5]
```diff
@@ -8,21 +8,60 @@
...
-use timely::progress::Antichain;
 use tracing::{debug, warn};
 
 const POLICIES: &[&str] = &[REFRESH_POLICY_NAME];
 
 const REFRESH_POLICY_NAME: &str = "refresh";
 
+#[derive(Clone, Debug)]
+pub struct SchedulingDecision {
+    on_off: bool,
```
[the remark this answers is not in the document]
Comment (ggevay) [reply src/adapter/src/coord/cluster_scheduling.rs]: Renamed to `cluster_on` and also added a comment. e5426b502072f4f99565c1928e74882540d327c1

File: src/adapter/src/catalog/transact.rs [hunk 6]
```diff
@@ -94,6 +95,7 @@ pub enum Op {
         name: String,
         config: ReplicaConfig,
         owner_id: RoleId,
+        scheduling_decision_reasons: Option<Vec<SchedulingDecisionReason>>,
```
[the remark this answers is not in the document]
Comment (ggevay) [reply src/adapter/src/catalog/transact.rs]: Done. 000a052af1b0530148bb8f39246a210e0e1557f2

File: src/adapter/src/catalog/transact.rs [hunk 7]
```diff
@@ -95,7 +97,7 @@ pub enum Op {
         name: String,
         config: ReplicaConfig,
         owner_id: RoleId,
-        scheduling_decision_reasons: Option<Vec<SchedulingDecisionReason>>,
+        scheduling_decisions_with_reasons: Option<SchedulingDecisionsWithReasonsV1>,
```
[the remark this answers is not in the document]
Comment (ggevay) [reply src/adapter/src/catalog/transact.rs]: > We definitely don't want to leak the audit_log types here though. Can you make a separate enum or struct that we convert into `SchedulingDecisionsWithReasonsV1`?

Actually, a `Vec<SchedulingDecision>` can serve this purpose, so I just moved the conversion to the audit log struct to happen later. So where we were passing around `SchedulingDecisionsWithReasonsV1` in non-audit log code, we are now passing around `Vec<SchedulingDecision>`, and only converting to `SchedulingDecisionsWithReasonsV1` at the last possible moment.

3212323e056c0c433875e99fca64c26d765d8b43

File: src/adapter/src/catalog/transact.rs [hunk 8]
```diff
@@ -106,7 +108,8 @@ pub enum Op {
         sub_component: Option<usize>,
         comment: Option<String>,
     },
-    DropObjects(Vec<ObjectId>),
+    /// SchedulingDecisionReason is forwarded to mz_audit_events.details (only for replica drops).
+    DropObjects(Vec<(ObjectId, Option<Vec<SchedulingDecisionReason>>)>),
```
[the remark this answers is not in the document]
Comment (ggevay) [reply src/adapter/src/catalog/transact.rs]: Yes, I'll try to do this refactor now.

[the remark this answers is not in the document]
Comment (ParkMyCar) [reply src/adapter/src/catalog/transact.rs]: Thank you! If it ends up being too large of a change or too tricky to make, don't worry about it we can always follow up with it in a later PR

**Review by ParkMyCar — APPROVED** [review]
Thanks for iterating @ggevay! w.r.t. the `DropObjects` refactor we can always follow up with that, as-is this looks good to me :)

**Comment by ggevay** [comment]
Almost done with the refactor, it's not too bad, and indeed makes the code nicer :)

**Comment by ParkMyCar** [comment]
> Almost done with the refactor, it's not too bad, and indeed makes the code nicer :)

Sounds great!!

File: src/adapter/src/catalog/transact.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (ggevay) [reply src/adapter/src/catalog/transact.rs]: Done: a01ecc8da87d8ab7b23b6ba1665a71bf135b11f3

**Comment by ggevay** [comment]
Done: a01ecc8da87d8ab7b23b6ba1665a71bf135b11f3

I've also moved moved `ReplicaCreateDropReason` to `transact.rs`: c5a79144e6f8c5756b04fd5c899da54d67edde89

**Comment by ggevay** [comment]
Kicked off a fresh Nightly: https://buildkite.com/materialize/nightly/builds/7928

**Pull request merged**